File: android/app/src/main/java/com/forrestlasiter/umbra/vpn/SinkholeStats.kt
package com.forrestlasiter.umbra.vpn

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/**
 * Process-wide live stats for the DNS sinkhole, so the UI can show what the
 * tunnel is actually doing (blocked vs forwarded) without binding to the service.
 * The service and the Activity run in the same process, so a shared StateFlow is
 * enough.
 */
object SinkholeStats {

    data class Snapshot(
        val active: Boolean = false,
        val domains: Int = 0,
        val blocked: Long = 0,
        val forwarded: Long = 0,
    )

    private val _flow = MutableStateFlow(Snapshot())
    val flow: StateFlow<Snapshot> = _flow.asStateFlow()

    fun started(domains: Int) { _flow.value = Snapshot(active = true, domains = domains) }

    fun update(blocked: Long, forwarded: Long) {
        _flow.value = _flow.value.copy(blocked = blocked, forwarded = forwarded)
    }

    fun stopped() { _flow.value = _flow.value.copy(active = false) }
}

File: ios/UmbraCore/Sources/UmbraCore/SinkholeStats.swift
import Foundation

/// Live DNS-sinkhole counters shared between the Packet Tunnel extension (which
/// writes them) and the app (which reads them). The two run in separate
/// processes, so they share an App Group `UserDefaults` suite. The suite is
/// injected, which also makes this unit-testable.
public struct SinkholeStats {
    public static let appGroup = "group.com.forrestlasiter.umbra"

    public struct Snapshot: Equatable {
        public let active: Bool
        public let domains: Int
        public let blocked: Int
        public let forwarded: Int
    }

    private let defaults: UserDefaults

    public init(defaults: UserDefaults) { self.defaults = defaults }

    /// The app-group-backed store (nil if the entitlement isn't present).
    public static func shared() -> SinkholeStats? {
        UserDefaults(suiteName: appGroup).map(SinkholeStats.init)
    }

    public func started(domains: Int) {
        defaults.set(true, forKey: Key.active)
        defaults.set(domains, forKey: Key.domains)
        defaults.set(0, forKey: Key.blocked)
        defaults.set(0, forKey: Key.forwarded)
    }

    public func recordBlocked() { defaults.set(defaults.integer(forKey: Key.blocked) + 1, forKey: Key.blocked) }
    public func recordForwarded() { defaults.set(defaults.integer(forKey: Key.forwarded) + 1, forKey: Key.forwarded) }
    public func stopped() { defaults.set(false, forKey: Key.active) }

    public func snapshot() -> Snapshot {
        Snapshot(active: defaults.bool(forKey: Key.active),
                 domains: defaults.integer(forKey: Key.domains),
                 blocked: defaults.integer(forKey: Key.blocked),
                 forwarded: defaults.integer(forKey: Key.forwarded))
    }

    private enum Key {
        static let active = "umbra.sink.active"
        static let domains = "umbra.sink.domains"
        static let blocked = "umbra.sink.blocked"
        static let forwarded = "umbra.sink.forwarded"
    }
}
